pkg/autoscaler/stats_scraper.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package autoscaler
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/knative/serving/pkg/apis/serving"
+	"github.com/knative/serving/pkg/reconciler"
+	"github.com/prometheus/common/expfmt"
+	"go.uber.org/zap"
+)
+
+// StatsScraper defines the interface for collecting Revision metrics
+type StatsScraper interface {
+	// Scrape scrapes the Revision queue metric endpoint then sends it as a
+	// StatMessage to the given channel.
+	Scrape(statsCh chan<- *StatMessage)
+}
+
+// cacheDisabledClient is a http client with cache disabled. It is shared by
+// every goruntime for a revision scraper.
+var cacheDisabledClient = &http.Client{
+	Transport: &http.Transport{
+		// Do not use the cached connection
+		DisableKeepAlives: true,
+	},
+	Timeout: 3 * time.Second,
+}
+
+// ServiceScraper scrapes Revision metrics via a K8S service by sampling. Which
+// pod to be picked up to serve the request is decided by K8S.
+type ServiceScraper struct {
+	httpClient *http.Client
+	url        string
+	metricKey  string
+	logger     *zap.SugaredLogger
+}
+
+// NewServiceScraper creates a new StatsScraper for the Revision which
+// the given Metric is responsible for.
+func NewServiceScraper(metric *Metric, logger *zap.SugaredLogger) (*ServiceScraper, error) {
+	return newServiceScraperWithClient(metric, logger, cacheDisabledClient)
+}
+
+func newServiceScraperWithClient(metric *Metric, logger *zap.SugaredLogger, httpClient *http.Client) (*ServiceScraper, error) {
```

Comment:
given there's just one public constructor, I think just putting this directly in the `NewServiceScraper` is totally fine.